typhos/cache.py

```diff
@@ -210,3 +225,112 @@ def get(self, obj):
             if desc is not None:
                 # In certain scenarios (such as testing) this might happen
                 self._new_description(obj, desc)
+
+
+# The default stale cached_path threshold time, in seconds:
+TYPHOS_PATH_CACHE_TIME = int(os.environ.get('TYPHOS_PATH_CACHE_TIME', '600'))
+
+
+class _CachedPath:
+    """
+    A wrapper around pathlib.Path to support repeated globbing
+
+    Parameters
+    ----------
+    path : pathlib.Path
+        The path to cache
+
+    Attributes
+    ----------
+    path : pathlib.Path
+        The underlying path
+    cache : list
+        The cache of filenames
+    _update_time : float
+        The last time the cache was updated
+    stale_threshold : float, optional
+        The time (in seconds) after which to update the path cache.  This
+        happens on the next glob, and not on a timer-basis.
+    """
+
+    def __init__(self, path, *, stale_threshold=TYPHOS_PATH_CACHE_TIME):
+        self.path = pathlib.Path(path)
+        self.cache = None
+        self._update_time = None
+        self.stale_threshold = stale_threshold
+
+    @classmethod
+    def from_path(cls, path, **kwargs):
+        """
+        Get a cached path, if not already cached.
+
+        Parameters
+        ----------
+        path : :class:`pathlib.Path` or :class:`_CachedPath`
+            The paths to cache, if not already cached.
+        """
+        if isinstance(path, (cls, _GlobalDisplayPathCache)):
+            # Already a _CachedPath
+            return path
+        return cls(path, **kwargs)
+
+    def __hash__(self):
+        # Keep the hash the same as the internal path for set()/dict() usage
+        return hash(self.path)
+
+    @property
+    def time_since_last_update(self):
+        """Time (in seconds) since the last update."""
+        if self._update_time is None:
+            return 0
+        return time.monotonic() - self._update_time
+
+    def update(self):
+        """Update the file list"""
+        self.cache = os.listdir(self.path)
+        self._update_time = time.monotonic()
+
+    def glob(self, pattern):
+        """Glob a pattern"""
+        if self.cache is None:
+            self.update()
+        elif self.time_since_last_update > self.stale_threshold:
+            self.update()
+
+        if any(c in pattern for c in '*?['):
+            # Convert from glob syntax -> regular expression
+            # And compile it for repeated usage.
+            regex = re.compile(fnmatch.translate(pattern))
+            for path in self.cache:
+                if regex.match(path):
+                    yield self.path / path
+        else:
+            # No globbing syntax: only check if file is in the list
+            if pattern in self.cache:
+                yield self.path / pattern
+
+
+class _GlobalDisplayPathCache:
+    """
+    A cache for all configured display paths
```

Comment:
Our new style guide says that docstrings should end in periods

Reply:
Resolved in 6e51a31be01f94cb31d6b0cbd58cb4e5157aece1